boot/boot.go

```diff
@@ -143,29 +142,77 @@ func applicable(s snap.PlaceInfo, t snap.Type, dev Device) bool {
 	return true
 }
 
-// InUse checks if the given name/revision is used in the
-// boot environment
-func InUse(name string, rev snap.Revision) bool {
-	bootloader, err := bootloader.Find("", nil)
-	if err != nil {
-		logger.Noticef("cannot get boot settings: %s", err)
-		return false
+// bootState exposes the boot state for a type of boot snap.
+type bootState interface {
+	// revisions retrieves the revisions of the current snap and
+	// the try snap (only the latter might not be set), and
+	// whether the snap is in "trying" state.
+	revisions() (snap, try_snap *NameAndRevision, trying bool, err error)
+}
+
+// bootStateFor finds the right bootState implementation of the given
+// snap type and Device, if applicable.
+func bootStateFor(typ snap.Type, dev Device) (s bootState, err error) {
+	if !dev.RunMode() {
+		return nil, fmt.Errorf("internal error: no boot state handling for ephemeral modes")
 	}
+	switch typ {
+	case snap.TypeOS, snap.TypeBase:
+		return newBootState16(snap.TypeBase), nil
+	case snap.TypeKernel:
+		return newBootState16(snap.TypeKernel), nil
+	default:
+		return nil, fmt.Errorf("internal error: no boot state handling for snap type %q", typ)
+	}
+}
 
-	bootVars, err := bootloader.GetBootVars("snap_kernel", "snap_try_kernel", "snap_core", "snap_try_core")
-	if err != nil {
-		logger.Noticef("cannot get boot vars: %s", err)
-		return false
+type InUseFunc func(name string, rev snap.Revision) bool
+
+func fixedInUse(inUse bool) InUseFunc {
+	return func(string, snap.Revision) bool {
+		return inUse
 	}
+}
 
-	snapFile := filepath.Base(snap.MountFile(name, rev))
-	for _, bootVar := range bootVars {
-		if bootVar == snapFile {
-			return true
-		}
+// InUse returns a checker for whether a given name/revision is used in the
+// boot environment for snaps of the relevant snap type.
+func InUse(typ snap.Type, dev Device) (InUseFunc, error) {
```

Comment:
I was wondering if we somehow could avoid the indirection of the InUseFunc - but it makes a lot of things nice so maybe something to explore once we refactored and implemented the new uc20 boot code.

Reply:
there is probably an approach were we make the bootStates explicit/exported in some form, but yes, something maybe to change/clean up later